compositor_common/src/validators.rs

```diff
@@ -98,4 +122,134 @@ impl SceneSpec {
 
         Ok(())
     }
+
+    fn validate_cycles(
+        &self,
+        registered_inputs: &HashSet<NodeId>,
+        transform_nodes: &HashMap<&NodeId, &TransformNodeSpec>,
+    ) -> Result<(), SpecValidationError> {
+        enum NodeState {
+            BeingVisited,
+            Visited,
+        }
+
+        let mut visited: HashMap<&NodeId, NodeState> = HashMap::new();
+
+        fn visit<'a>(
+            node: &'a NodeId,
+            transform_nodes: &'a HashMap<&NodeId, &TransformNodeSpec>,
+            inputs: &HashSet<NodeId>,
+            visited: &mut HashMap<&'a NodeId, NodeState>,
+        ) -> Result<(), SpecValidationError> {
+            if let Some(_input) = inputs.get(node) {
+                return Ok(());
+            }
+
+            match visited.get(node) {
+                Some(NodeState::BeingVisited) => return Err(SpecValidationError::CycleDetected),
+                Some(NodeState::Visited) => return Ok(()),
+                None => {}
+            }
+
+            visited.insert(node, NodeState::BeingVisited);
+
+            for child in &transform_nodes.get(node).unwrap().input_pads {
+                visit(child, transform_nodes, inputs, visited)?;
+            }
+
+            visited.insert(node, NodeState::Visited);
+
+            Ok(())
+        }
+
+        for output in &self.outputs {
+            visit(
+                &output.input_pad,
+                transform_nodes,
+                registered_inputs,
+                &mut visited,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn validate_node_ids_uniqueness<'a, I: Iterator<Item = &'a NodeId>>(
+        &self,
+        defined_node_ids: I,
+    ) -> Result<(), SpecValidationError> {
+        let mut nodes_ids: HashSet<&NodeId> = HashSet::new();
+
+        for node_id in defined_node_ids {
+            if nodes_ids.contains(&node_id) {
+                return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
+            }
+            nodes_ids.insert(node_id);
```

Comment:
```suggestion
            if !nodes_ids.insert(node_id) {
                return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
            }
```

`insert` returns whether the insertion actually happened or the value was already in the set